aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -297,19 +297,34 @@ pub struct AggregatorData {
 
 impl AggregatorData {
     /// Returns a mutable reference to an aggregator with `id` and a `limit`.
-    /// If transaction that is currently executing did not initilize it), a new
-    /// aggregator instance is created, with a zero-initialized value and in a
-    /// delta state.
+    /// If transaction that is currently executing did not initilize it, a new aggregator instance is created.
+    /// The state of the new aggregator instance depends on the `aggregator_enabled` flag.
+    /// If the `aggregator_enabled` flag is true, the new aggregator instance
+    /// is initialized with zero and in a delta state.
+    /// If the `aggregator_enabled` flag is false, the new aggregator instance
+    /// is set initialized with its latest value and in a Data state.
     /// Note: when we say "aggregator instance" here we refer to Rust struct and
     /// not to the Move aggregator.
-    pub fn get_aggregator(&mut self, id: AggregatorID, limit: u128) -> &mut Aggregator {
+    pub fn get_aggregator(
+        &mut self,
+        id: AggregatorID,
+        limit: u128,
+        resolver: &dyn TableResolver,
+        aggregator_enabled: bool,
+    ) -> &mut Aggregator {
         self.aggregators.entry(id).or_insert_with(|| Aggregator {
             value: 0,
             state: AggregatorState::PositiveDelta,
             limit,
             history: Some(History::new()),
         });
-        self.aggregators.get_mut(&id).unwrap()
+
+        let aggregator = self.aggregators.get_mut(&id).unwrap();
+        if !aggregator_enabled {
```

Comment:
nit:
```rust
if !aggregator_enabled {
    aggregator.read_and_materialize(resolver, &id).unwrap_or(0);
}
```
should just work?

Reply:
@georgemitenkov @gelash As you mentioned, I thought `read_and_materialize` might fail if there is no data corresponding to the aggregator. I thought using `unwrap_or` handles this case gracefully. So, is your final recommendation to keep things as it is? or do you want me to change assuming that `read_and_materialize` never fails?

Reply:
I think George's proposal (if I understand right) is the best: change the output type of get_aggregator function to return PartialVMResult, do `aggregator.read_and_materialize(resolver, &id)?; ` here instead of lines 324-325, and adjust the callers of get_aggregator to also do ?. + adjusting the tests / making sure this behavior is tested.

The reason it's better than unwrap_or - the empty case can only happen during incorrect speculative executions, so it's better to finish execution (txn anyway will be re-executed) asap and save work/time. Also from code prospective, assigning value, then state separately is error prone (as George pointed out you'd also likely need to assign history, and in general if things change it's hard to remember to also adjust this place).

@georgemitenkov does this match to what you proposed?

Reply:
Yes :)

Reply:
I made the `get_aggregator` method return `PartialVMResult` and changed the unit tests accordingly. But one unit test in `context.rs` is failing. I didn't completely understand why it's failing.

Reply:
Does `test_into_change_set_aggregator_disabled` fail? Or what unit test?

That one fails because you don't have any value in storage: you call 
```
aggregator_data
            .get_aggregator(
                aggregator_id_for_test(700),
                700,
                context.resolver,
                aggregator_enabled,
            )
            .unwrap()
            .add(200)
            .unwrap();
```
which panics on first `unwrap()` I assume. This is because your storage is empty (that's why it's called `EmptyStorage` in a test and there are no values in a table). Storage here is just a `HashMap<StateKey, Vec<u8>>` so you can place there serialized value of `200` for state key corresponding to your aggregator, which should fix the problem for that `get_aggregator`. Does this help?

Reply:
Also we should have a test that we error on having nothing in storage (basically what your test does right now bu instead of panic we should `assert_matches!` on error.

Reply:
But the fact that it failing makes sense, because you made aggregator creation optional. Here
```
if !aggregator_enabled {
    aggregator_data.create_new_aggregator(aggregator_id_for_test(500), 500);
    // ...
}
```
So of course storage has `None`, and test fails.